File: electron/main/managed/git/paths.ts
import { randomUUID } from 'node:crypto'
import { lstat, mkdir, readFile, realpath, rmdir, unlink, writeFile } from 'node:fs/promises'
import nodePath from 'node:path'
import { GitRuntimeError, gitRuntimeFailure } from './errors'
import type { ManagedGitInstallationPaths } from './types'

const INSTALLATION_ID = /^[a-z][a-z0-9_-]{2,127}$/

/** Builds the fixed, launcher-owned Git layout below an already-validated Harness namespace. */
export function createManagedGitInstallationPaths(
  namespacePath: string,
  installationId: string
): ManagedGitInstallationPaths {
  assertCanonicalAbsolutePath(namespacePath, 'Managed Harness namespace')
  if (!INSTALLATION_ID.test(installationId)) {
    throw new GitRuntimeError('git.managed_path_invalid', 'Managed installation id is invalid.')
  }
  const installationPath = nodePath.join(
    namespacePath,
    'dsh-launcher',
    'managed-installations',
    installationId
  )
  const paths: ManagedGitInstallationPaths = {
    installationId,
    namespacePath,
    installationPath,
    mirrorPath: nodePath.join(installationPath, 'mirror.git'),
    worktreesPath: nodePath.join(installationPath, 'worktrees'),
    stagingPath: nodePath.join(installationPath, 'staging'),
    lockPath: nodePath.join(installationPath, 'operation.lock')
  }
  for (const candidate of [
    paths.installationPath,
    paths.mirrorPath,
    paths.worktreesPath,
    paths.stagingPath,
    paths.lockPath
  ]) {
    assertContainedManagedPath(paths, candidate)
  }
  return paths
}

/** Creates only the fixed launcher-owned parent directories, rejecting symlinks and unexpected files. */
export async function ensureManagedGitInstallationDirectories(
  paths: ManagedGitInstallationPaths
): Promise<void> {
  await assertCanonicalExistingDirectory(paths.namespacePath, 'Managed Harness namespace')
  await ensureDirectoryBelow(
    paths.namespacePath,
    nodePath.join(paths.namespacePath, 'dsh-launcher')
  )
  await ensureDirectoryBelow(
    paths.namespacePath,
    nodePath.join(paths.namespacePath, 'dsh-launcher', 'managed-installations')
  )
  await ensureDirectoryBelow(paths.namespacePath, paths.installationPath)
  await ensureDirectoryBelow(paths.namespacePath, paths.worktreesPath)
  await ensureDirectoryBelow(paths.namespacePath, paths.stagingPath)
}

/** Proves that an existing target is a direct, non-symlink child of the managed installation. */
export async function assertManagedGitTarget(
  paths: ManagedGitInstallationPaths,
  target: string,
  subject: string
): Promise<void> {
  assertContainedManagedPath(paths, target)
  let metadata
  try {
    metadata = await lstat(target)
  } catch (error) {
    throw gitRuntimeFailure('git.managed_path_invalid', `${subject} is unavailable.`, error, {
      installationId: paths.installationId
    })
  }
  if (metadata.isSymbolicLink()) {
    throw new GitRuntimeError(
      'git.managed_path_escape',
      `${subject} must not be a symbolic link.`,
      {
        installationId: paths.installationId
      }
    )
  }
  let canonicalTarget: string
  try {
    canonicalTarget = await realpath(target)
  } catch (error) {
    throw gitRuntimeFailure(
      'git.managed_path_invalid',
      `${subject} cannot be canonicalized.`,
      error,
      {
        installationId: paths.installationId
      }
    )
  }
  if (canonicalTarget !== target || !isStrictlyInside(paths.namespacePath, canonicalTarget)) {
    throw new GitRuntimeError(
      'git.managed_path_escape',
      `${subject} escapes the managed Harness namespace.`,
      {
        installationId: paths.installationId
      }
    )
  }
}

/** Acquires a non-stealable, launcher-owned operation lock for one managed installation. */
export async function acquireManagedGitOperationLock(
  paths: ManagedGitInstallationPaths
): Promise<ManagedGitOperationLock> {
  await ensureManagedGitInstallationDirectories(paths)
  try {
    await mkdir(paths.lockPath, { mode: 0o700 })
  } catch (error) {
    if (isNodeCode(error, 'EEXIST')) {
      throw new GitRuntimeError(
        'git.operation_locked',
        'A managed Git operation is already in progress.',
        {
          installationId: paths.installationId
        }
      )
    }
    throw gitRuntimeFailure(
      'git.operation_failed',
      'Managed Git operation lock could not be created.',
      error,
      {
        installationId: paths.installationId
      }
    )
  }

  const ownerId = randomUUID()
  const ownerPath = nodePath.join(paths.lockPath, 'owner.json')
  try {
    await writeFile(ownerPath, JSON.stringify({ ownerId }) + '\n', {
      encoding: 'utf8',
      flag: 'wx',
      mode: 0o600
    })
  } catch (error) {
    await rmdir(paths.lockPath).catch(() => undefined)
    throw gitRuntimeFailure(
      'git.operation_failed',
      'Managed Git operation lock could not record its owner.',
      error,
      {
        installationId: paths.installationId
      }
    )
  }
  return new ManagedGitOperationLock(paths, ownerId, ownerPath)
}

/** Owns release of exactly the lock directory created by this process. */
export class ManagedGitOperationLock {
  readonly #paths: ManagedGitInstallationPaths
  readonly #ownerId: string
  readonly #ownerPath: string
  #released = false

  constructor(paths: ManagedGitInstallationPaths, ownerId: string, ownerPath: string) {
    this.#paths = paths
    this.#ownerId = ownerId
    this.#ownerPath = ownerPath
  }

  /** Releases the lock only when its owner record still belongs to this instance. */
  async release(): Promise<void> {
    if (this.#released) {
      throw new GitRuntimeError(
        'git.operation_lock_lost',
        'Managed Git operation lock was already released.',
        {
          installationId: this.#paths.installationId
        }
      )
    }
    let text: string
    try {
      text = await readFile(this.#ownerPath, 'utf8')
    } catch (error) {
      throw gitRuntimeFailure(
        'git.operation_lock_lost',
        'Managed Git operation lock owner record is unavailable.',
        error,
        {
          installationId: this.#paths.installationId
        }
      )
    }
    if (text !== `${JSON.stringify({ ownerId: this.#ownerId })}\n`) {
      throw new GitRuntimeError(
        'git.operation_lock_lost',
        'Managed Git operation lock owner changed.',
        {
          installationId: this.#paths.installationId
        }
      )
    }
    try {
      await unlink(this.#ownerPath)
      await rmdir(this.#paths.lockPath)
    } catch (error) {
      throw gitRuntimeFailure(
        'git.operation_lock_lost',
        'Managed Git operation lock could not be released.',
        error,
        {
          installationId: this.#paths.installationId
        }
      )
    }
    this.#released = true
  }
}

/** Returns an exact detached-worktree path only for a full resolved commit SHA. */
export function managedWorktreePath(paths: ManagedGitInstallationPaths, commit: string): string {
  if (!/^[0-9a-f]{40}$/.test(commit)) {
    throw new GitRuntimeError(
      'git.ref_invalid',
      'Managed worktree requires a full lowercase commit SHA.'
    )
  }
  const target = nodePath.join(paths.worktreesPath, commit)
  assertContainedManagedPath(paths, target)
  return target
}

/** Asserts that an arbitrary supplied path remains below the authoritative namespace. */
export function assertContainedManagedPath(
  paths: ManagedGitInstallationPaths,
  target: string
): void {
  assertCanonicalAbsolutePath(target, 'Managed Git path')
  if (!isStrictlyInside(paths.namespacePath, target)) {
    throw new GitRuntimeError(
      'git.managed_path_escape',
      'Managed Git path escapes the Harness namespace.',
      {
        installationId: paths.installationId
      }
    )
  }
}

async function ensureDirectoryBelow(namespacePath: string, target: string): Promise<void> {
  if (!isStrictlyInside(namespacePath, target)) {
    throw new GitRuntimeError(
      'git.managed_path_escape',
      'Managed Git directory escapes the Harness namespace.'
    )
  }
  const relative = nodePath.relative(namespacePath, target)
  const segments = relative.split(nodePath.sep)
  let cursor = namespacePath
  for (const segment of segments) {
    cursor = nodePath.join(cursor, segment)
    try {
      const metadata = await lstat(cursor)
      if (!metadata.isDirectory() || metadata.isSymbolicLink()) {
        throw new GitRuntimeError(
          'git.managed_path_escape',
          'Managed Git directory is not a direct directory.'
        )
      }
    } catch (error) {
      if (error instanceof GitRuntimeError) throw error
      if (!isNodeCode(error, 'ENOENT')) {
        throw gitRuntimeFailure(
          'git.operation_failed',
          'Managed Git directory could not be inspected.',
          error
        )
      }
      try {
        await mkdir(cursor, { mode: 0o700 })
      } catch (createError) {
        throw gitRuntimeFailure(
          'git.operation_failed',
          'Managed Git directory could not be created.',
          createError
        )
      }
      try {
        const created = await lstat(cursor)
        if (!created.isDirectory() || created.isSymbolicLink()) {
          throw new GitRuntimeError(
            'git.managed_path_escape',
            'Managed Git directory changed during creation.'
          )
        }
      } catch (verifyError) {
        if (verifyError instanceof GitRuntimeError) throw verifyError
        throw gitRuntimeFailure(
          'git.operation_failed',
          'Managed Git directory could not be verified after creation.',
          verifyError
        )
      }
    }
  }
}

async function assertCanonicalExistingDirectory(path: string, subject: string): Promise<void> {
  assertCanonicalAbsolutePath(path, subject)
  let metadata
  let canonicalPath: string
  try {
    metadata = await lstat(path)
    canonicalPath = await realpath(path)
  } catch (error) {
    throw gitRuntimeFailure('git.managed_path_invalid', `${subject} is unavailable.`, error)
  }
  if (!metadata.isDirectory() || metadata.isSymbolicLink() || canonicalPath !== path) {
    throw new GitRuntimeError(
      'git.managed_path_invalid',
      `${subject} must be an existing canonical directory.`
    )
  }
}

function assertCanonicalAbsolutePath(value: unknown, subject: string): asserts value is string {
  if (
    typeof value !== 'string' ||
    !value ||
    value.includes('\u0000') ||
    !nodePath.isAbsolute(value)
  ) {
    throw new GitRuntimeError('git.managed_path_invalid', `${subject} must be an absolute path.`)
  }
  if (nodePath.normalize(value) !== value || nodePath.parse(value).root === value) {
    throw new GitRuntimeError(
      'git.managed_path_invalid',
      `${subject} must be canonical and non-root.`
    )
  }
}

function isStrictlyInside(parent: string, child: string): boolean {
  const relative = nodePath.relative(parent, child)
  return relative !== '' && !relative.startsWith('..') && !nodePath.isAbsolute(relative)
}

function isNodeCode(value: unknown, expected: string): boolean {
  return Boolean(value && typeof value === 'object' && 'code' in value && value.code === expected)
}

File: networking/internal/gitcheckout/paths.go
package gitcheckout

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// installationIDPattern is the only shape a managed installation id may have: it
// becomes a directory name below the namespace, so it cannot be a path.
var installationIDPattern = regexp.MustCompile(`^[a-z][a-z0-9_-]{2,127}$`)

// InstallationPaths is the fixed, launcher-owned layout of one managed
// installation.
type InstallationPaths struct {
	InstallationID   string `json:"installationId"`
	NamespacePath    string `json:"namespacePath"`
	InstallationPath string `json:"installationPath"`
	MirrorPath       string `json:"mirrorPath"`
	WorktreesPath    string `json:"worktreesPath"`
	StagingPath      string `json:"stagingPath"`
	LockPath         string `json:"lockPath"`
}

// NewInstallationPaths builds the layout below an already-validated Harness
// namespace. Every path it derives is proved to stay inside that namespace, so a
// later caller cannot be handed a path that is already an escape.
func NewInstallationPaths(namespacePath, installationID string) (InstallationPaths, error) {
	if err := AssertCanonicalAbsolutePath(namespacePath, "Managed Harness namespace", ErrManagedPathInvalid); err != nil {
		return InstallationPaths{}, err
	}
	if !installationIDPattern.MatchString(installationID) {
		return InstallationPaths{}, fmt.Errorf("%w: Managed installation id is invalid.", ErrManagedPathInvalid)
	}
	installationPath := filepath.Join(namespacePath, "dsh-launcher", "managed-installations", installationID)
	paths := InstallationPaths{
		InstallationID:   installationID,
		NamespacePath:    namespacePath,
		InstallationPath: installationPath,
		MirrorPath:       filepath.Join(installationPath, "mirror.git"),
		WorktreesPath:    filepath.Join(installationPath, "worktrees"),
		StagingPath:      filepath.Join(installationPath, "staging"),
		LockPath:         filepath.Join(installationPath, "operation.lock"),
	}
	for _, candidate := range []string{
		paths.InstallationPath,
		paths.MirrorPath,
		paths.WorktreesPath,
		paths.StagingPath,
		paths.LockPath,
	} {
		if err := AssertContainedManagedPath(paths, candidate); err != nil {
			return InstallationPaths{}, err
		}
	}
	return paths, nil
}

// SameRegisteredPath compares a git-reported path with its registered spelling.
// Git for Windows reports rev-parse paths with forward slashes and Windows path
// matching is case-insensitive; both spellings name the directory the launcher
// registered, so neither difference is a mismatch.
func SameRegisteredPath(observed, registered string) bool {
	normalizedObserved := filepath.Clean(observed)
	normalizedRegistered := filepath.Clean(registered)
	if isWindows() {
		return strings.EqualFold(normalizedObserved, normalizedRegistered)
	}
	return normalizedObserved == normalizedRegistered
}

// EnsureInstallationDirectories creates only the fixed launcher-owned parent
// directories, refusing a symlink or an unexpected file at any level. The mirror
// itself is not created here: creating it is what the mirror rules do.
func EnsureInstallationDirectories(paths InstallationPaths) error {
	if err := AssertCanonicalExistingDirectory(paths.NamespacePath, "Managed Harness namespace"); err != nil {
		return err
	}
	for _, target := range []string{
		filepath.Join(paths.NamespacePath, "dsh-launcher"),
		filepath.Join(paths.NamespacePath, "dsh-launcher", "managed-installations"),
		paths.InstallationPath,
		paths.WorktreesPath,
		paths.StagingPath,
	} {
		if err := ensureDirectoryBelow(paths.NamespacePath, target); err != nil {
			return err
		}
	}
	return nil
}

// AssertManagedTarget proves that an existing target is a direct, non-symlink
// child of the managed namespace: it is contained, it is not a link, its realpath
// is its own spelling, and that realpath is still inside the namespace.
func AssertManagedTarget(paths InstallationPaths, target, subject string) error {
	if err := AssertContainedManagedPath(paths, target); err != nil {
		return err
	}
	metadata, err := os.Lstat(target)
	if err != nil {
		return fmt.Errorf("%w: %s is unavailable.", ErrManagedPathInvalid, subject)
	}
	if metadata.Mode()&os.ModeSymlink != 0 {
		return fmt.Errorf("%w: %s must not be a symbolic link.", ErrManagedPathEscape, subject)
	}
	canonicalTarget, err := filepath.EvalSymlinks(target)
	if err != nil {
		return fmt.Errorf("%w: %s cannot be canonicalized.", ErrManagedPathInvalid, subject)
	}
	if canonicalTarget != target || !isStrictlyInside(paths.NamespacePath, canonicalTarget) {
		return fmt.Errorf("%w: %s escapes the managed Harness namespace.", ErrManagedPathEscape, subject)
	}
	return nil
}

// AssertContainedManagedPath refuses any path that is not a canonical absolute
// path strictly below the authoritative namespace.
func AssertContainedManagedPath(paths InstallationPaths, target string) error {
	if err := AssertCanonicalAbsolutePath(target, "Managed Git path", ErrManagedPathInvalid); err != nil {
		return err
	}
	if !isStrictlyInside(paths.NamespacePath, target) {
		return fmt.Errorf("%w: Managed Git path escapes the Harness namespace.", ErrManagedPathEscape)
	}
	return nil
}

// ManagedWorktreePath returns the detached-worktree path of one full commit SHA.
// An abbreviated object id would name a directory that is not this commit, so it
// is refused rather than expanded.
func ManagedWorktreePath(paths InstallationPaths, commit string) (string, error) {
	sha, err := ParseCommitSHA(commit)
	if err != nil {
		return "", err
	}
	target := filepath.Join(paths.WorktreesPath, sha)
	if err := AssertContainedManagedPath(paths, target); err != nil {
		return "", err
	}
	return target, nil
}

// OperationLock is the launcher's own non-stealable lock for one managed
// installation. It is a directory, so creating it is atomic on every platform
// this launcher runs on, and it records its owner so a release cannot remove a
// lock that was taken over by someone else.
type OperationLock struct {
	paths     InstallationPaths
	ownerID   string
	ownerPath string
	released  bool
}

// AcquireOperationLock takes the lock for one installation.
func AcquireOperationLock(paths InstallationPaths) (*OperationLock, error) {
	if err := EnsureInstallationDirectories(paths); err != nil {
		return nil, err
	}
	if err := os.Mkdir(paths.LockPath, 0o700); err != nil {
		if os.IsExist(err) {
			return nil, fmt.Errorf("%w: A managed Git operation is already in progress.", ErrOperationLocked)
		}
		return nil, fmt.Errorf("%w: Managed Git operation lock could not be created.", ErrOperationFailed)
	}
	ownerID, err := newOwnerID()
	if err != nil {
		_ = os.Remove(paths.LockPath)
		return nil, fmt.Errorf("%w: Managed Git operation lock could not record its owner.", ErrOperationFailed)
	}
	ownerPath := filepath.Join(paths.LockPath, "owner.json")
	file, err := os.OpenFile(ownerPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
	if err == nil {
		_, err = file.WriteString(ownerRecord(ownerID))
	}
	if closeErr := closeIfOpen(file); err == nil {
		err = closeErr
	}
	if err != nil {
		_ = os.Remove(paths.LockPath)
		return nil, fmt.Errorf("%w: Managed Git operation lock could not record its owner.", ErrOperationFailed)
	}
	return &OperationLock{paths: paths, ownerID: ownerID, ownerPath: ownerPath}, nil
}

// Release removes exactly the lock directory this instance created, and only
// while its owner record still names this instance.
func (lock *OperationLock) Release() error {
	if lock.released {
		return fmt.Errorf("%w: Managed Git operation lock was already released.", ErrOperationLockLost)
	}
	text, err := os.ReadFile(lock.ownerPath)
	if err != nil {
		return fmt.Errorf("%w: Managed Git operation lock owner record is unavailable.", ErrOperationLockLost)
	}
	if string(text) != ownerRecord(lock.ownerID) {
		return fmt.Errorf("%w: Managed Git operation lock owner changed.", ErrOperationLockLost)
	}
	if err = os.Remove(lock.ownerPath); err == nil {
		err = os.Remove(lock.paths.LockPath)
	}
	if err != nil {
		return fmt.Errorf("%w: Managed Git operation lock could not be released.", ErrOperationLockLost)
	}
	lock.released = true
	return nil
}

// ownerRecord is the exact owner document: the same bytes the shell wrote, so a
// lock taken by either implementation is recognized by the other.
// ownerRecord is the exact owner document: the same bytes the shell wrote, so a
// lock taken by either implementation is recognized by the other.
func ownerRecord(ownerID string) string {
	return fmt.Sprintf("{\"ownerId\":%q}\n", ownerID)
}

func newOwnerID() (string, error) {
	value := make([]byte, 16)
	if _, err := rand.Read(value); err != nil {
		return "", err
	}
	return hex.EncodeToString(value), nil
}

func closeIfOpen(file *os.File) error {
	if file == nil {
		return nil
	}
	return file.Close()
}

// ensureDirectoryBelow walks one path below the namespace, creating what is
// missing and refusing anything that is not a direct directory: a symlink at any
// level would let the rest of the walk leave the namespace.
func ensureDirectoryBelow(namespacePath, target string) error {
	if !isStrictlyInside(namespacePath, target) {
		return fmt.Errorf("%w: Managed Git directory escapes the Harness namespace.", ErrManagedPathEscape)
	}
	relative, err := filepath.Rel(namespacePath, target)
	if err != nil {
		return fmt.Errorf("%w: Managed Git directory escapes the Harness namespace.", ErrManagedPathEscape)
	}
	cursor := namespacePath
	for _, segment := range strings.Split(relative, string(filepath.Separator)) {
		cursor = filepath.Join(cursor, segment)
		metadata, err := os.Lstat(cursor)
		switch {
		case err == nil:
			if !metadata.IsDir() || metadata.Mode()&os.ModeSymlink != 0 {
				return fmt.Errorf("%w: Managed Git directory is not a direct directory.", ErrManagedPathEscape)
			}
		case !os.IsNotExist(err):
			return fmt.Errorf("%w: Managed Git directory could not be inspected.", ErrOperationFailed)
		default:
			if err = os.Mkdir(cursor, 0o700); err != nil {
				return fmt.Errorf("%w: Managed Git directory could not be created.", ErrOperationFailed)
			}
			created, verifyErr := os.Lstat(cursor)
			if verifyErr != nil || !created.IsDir() || created.Mode()&os.ModeSymlink != 0 {
				return fmt.Errorf("%w: Managed Git directory changed during creation.", ErrManagedPathEscape)
			}
		}
	}
	return nil
}

func AssertCanonicalExistingDirectory(path, subject string) error {
	if err := AssertCanonicalAbsolutePath(path, subject, ErrManagedPathInvalid); err != nil {
		return err
	}
	metadata, err := os.Lstat(path)
	if err != nil {
		return fmt.Errorf("%w: %s is unavailable.", ErrManagedPathInvalid, subject)
	}
	canonicalPath, err := filepath.EvalSymlinks(path)
	if err != nil {
		return fmt.Errorf("%w: %s is unavailable.", ErrManagedPathInvalid, subject)
	}
	if !metadata.IsDir() || metadata.Mode()&os.ModeSymlink != 0 || canonicalPath != path {
		return fmt.Errorf("%w: %s must be an existing canonical directory.", ErrManagedPathInvalid, subject)
	}
	return nil
}

// isStrictlyInside reports whether child is below parent and is not parent. The
// test is a path element comparison rather than a prefix comparison, so a sibling
// whose name merely begins with two dots is not mistaken for an escape.
func isStrictlyInside(parent, child string) bool {
	relative, err := filepath.Rel(parent, child)
	if err != nil {
		return false
	}
	return relative != "." && relative != ".." &&
		!strings.HasPrefix(relative, ".."+string(filepath.Separator)) &&
		!filepath.IsAbs(relative)
}
